### ISSUE TITLE
write unit test

### ISSUE THREAD

**Issue by gorazdko** [BlockchainCommons/sweeptool-cli#1]
## Abstract

Write more unit tests for UR serialization and deserialization.

First, there are small changes that need to be done to the UR test vectors, see https://github.com/BlockchainCommons/Research/issues/83 To correct these changes open a PR in Research repo. After approved, write the unit tests for those test vectors in this repo.

You'll need the following tool:
* https://github.com/BlockchainCommons/bytewords-cli
Some general knowledge about:
* URs: https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md
 * bytewords: https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-012-bytewords.md

**Comment by nochiel** [comment]
@gorazdko @wolfmcnally I'd like to start working on this.

**Comment by gorazdko** [comment]
Great. Only the test vector that has an incorrect depth should be fixed and a unit test written for that here. ("missing a parent fingerprint" test case is actually correct )

To give a little more context:  in https://github.com/BlockchainCommons/Research/issues/83 you can see depth is 4 ( `04 ; depth 4`) which cannot be correctly reconstructed from the components length (which is 3, i.e. `44, true, 0, true, 0, true ; 44'/0'/0'`). Therefore depth must be explicitly present in cbor notation.

**Comment by nochiel** [comment]
> To give a little more context: in [BlockchainCommons/Research#83](https://github.com/BlockchainCommons/Research/issues/83) you can see depth is 4 ( `04 ; depth 4`) which cannot be correctly reconstructed from the components length (which is 3, i.e. `44, true, 0, true, 0, true ; 44'/0'/0'`). Therefore depth must be explicitly present in cbor notation.

:grin: A dumb question for which I need some clarification:

Am I understanding this case correctly?
- Given `pkh([d34db33f/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/1/*)
`
- `m/44'/0'/0'` => depth 2  from the `master` with the given fingerprint.
- the given `xpub` is derived at the 0 index at depth 2. However, it is not possible to derive private keys up to depth 2 using this descriptor because each of the derivation steps has been hardened.
- Therefore we consider the given `xpub`, which is at depth 2, as our 0th step for the purpose of using this descriptor to derive scripts.
- Then, `xpub.../1/*` is the set of all keys at depth 1 from the given `xpub` ( == depth 4 from the `master` with the given fingerprint).

> [...] you can see depth is 4 ( `04 ; depth 4`) which cannot be correctly reconstructed from the components length.

Given the components: 
```cbor
6: 304({ ; origin: crypto-keypath
			1: [ ; components
				44, true, 0, true, 0, true ; 44'/0'/0'
			],
			2: 3545084735 ; origin-fingerprint
		}),
		7: 304({ ; children: crypto-keypath
			1: [ ; components
				1, false, [], false ; 1/*
			]
		}),
```

Isn't the depth the `sum(numbers of crypto-keypath components) - 1`? Doesn't this mean that we can in fact correctly reconstruct depth from the components' length?

**Comment by gorazdko** [comment]
> Therefore, you are saying depth is a required field (ref. cddl for key path) when components are present in crypto-keypath

No. I'm saying depth is required only when the components are not present or when the components are present and depth cannot be reconstructed from it. The latter is what the problem is here. This happens because we can come up with an xpub and derive a child from it. We can consider that child as a `new xpub` from which we can derive new children. If I hand you over that `new xpub` the information about depth cannot be restored from the component's array anymore.

> A test would check that (sum(length(components)) \forall components) - 1 == depth

try to apply that formula to other test vectors and you'll see it doesn't work

**Comment by nochiel** [comment]
> > Therefore, you are saying depth is a required field (ref. cddl for key path) when components are present in crypto-keypath
> 
> No. I'm saying depth is required only when the components are not present or when the components are present and depth cannot be reconstructed from it. The latter is what the problem is here.

Got it! Thank you!
